### PR TITLE
Add Category model

### PR DESCRIPTION
**Pull request by Stelmaschyk**

### CODE REVIEW

File: src/main/java/mate/academy/bookstore/config/SecurityConfig.java [hunk 1]
```diff
@@ -42,7 +42,10 @@ public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Excepti
                         "/error",
                         "/api/books/search",
                         "/api/books",
-                        "/api/books/{id}")
+                        "/api/books/{id}",
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/config/SecurityConfig.java]: let's refactor
```suggestion
                       .requestMatchers(
                                        antMatcher("/auth/**"),
                                        antMatcher("/swagger-ui/**"),
                                        antMatcher("/v3/api-docs/**"))
                                .permitAll()
```

File: src/main/java/mate/academy/bookstore/controller/CategoryController.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+@Tag(name = "Category", description = "Endpoints for managing categories")
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/api/categories")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/controller/CategoryController.java]: add s`erver.servlet.context-path=/api` to application.properties

```suggestion
@RequestMapping("/categories")
```

File: src/main/java/mate/academy/bookstore/controller/CategoryController.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+public class CategoryController {
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/controller/CategoryController.java]: remove empty line after class declaration
```suggestion
```

File: src/main/java/mate/academy/bookstore/controller/CategoryController.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+                schema = @Schema(implementation = CategoryResponseDto.class)
+            )
+    )
+    public CategoryResponseDto createCategory(@RequestBody @Valid CategoryRequestDto requestDto) {
+        return categoryService.save(requestDto);
+    }
+
+    @Operation(summary = "Get all available categories with pagination")
+    @GetMapping
+    public List<CategoryResponseDto> getAllCategories(Pageable pageable) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/controller/CategoryController.java]: it's better to return Page
```suggestion
    public Page<CategoryResponseDto> getAllCategories(Pageable pageable) {
```

File: src/main/java/mate/academy/bookstore/dto/book/CreateBookRequestDto.java [hunk 5]
```diff
@@ -23,4 +24,5 @@ public class CreateBookRequestDto {
     @Size(max = 150)
     private String description;
     private String coverImage;
+    private Set<Long> categoryIds;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/dto/book/CreateBookRequestDto.java:29]:
```suggestion
    @NotEmpty
    private Set<Long> categoryIds;
```

File: src/main/java/mate/academy/bookstore/dto/book/UpdateBookRequestDto.java [hunk 6]
```diff
@@ -23,4 +24,5 @@ public class UpdateBookRequestDto {
     @Size(max = 100)
     private String description;
     private String coverImage;
+    private Set<Long> categoryIds;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/dto/book/UpdateBookRequestDto.java:29]: same here

File: src/main/java/mate/academy/bookstore/model/Book.java [hunk 7]
```diff
@@ -34,6 +42,11 @@ public class Book {
     private String coverImage;
     @Column(nullable = false)
     private boolean isDeleted = false;
+    @ManyToMany(fetch = FetchType.LAZY)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/Book.java]:
```suggestion
    @ManyToMany
```

File: src/main/java/mate/academy/bookstore/model/Category.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private long id;
+    @Column(nullable = false)
+    private String name;
+    private String description;
+    @ToStringExclude
+    @EqualsAndHashCode.Exclude
+    @ManyToMany
+    private Set<Book> books = new HashSet<Book>();
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/Category.java]:
```suggestion
```

File: src/main/java/mate/academy/bookstore/model/Category.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+@Setter
+@SQLDelete(sql = "UPDATE categories SET is_deleted = true WHERE id=?")
+@SQLRestriction("is_deleted=false")
+@Table(name = "categories")
+@NoArgsConstructor
+public class Category {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private long id;
+    @Column(nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/Category.java]:
```suggestion
    @Column(nullable = false, unique = true)
```

File: src/main/java/mate/academy/bookstore/service/CategoryService.java [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+import java.util.List;
+import mate.academy.bookstore.dto.book.BookDto;
+import mate.academy.bookstore.dto.category.CategoryRequestDto;
+import mate.academy.bookstore.dto.category.CategoryResponseDto;
+import mate.academy.bookstore.dto.category.UpdateCategoryRequestDto;
+import org.springframework.data.domain.Pageable;
+import org.springframework.stereotype.Service;
+
+@Service
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/CategoryService.java]:
```suggestion
```

File: src/main/java/mate/academy/bookstore/service/CategoryServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    @Override
+    public CategoryResponseDto getById(Long id) {
+        Category category = categoryRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("Can't find category by id " + id));
+        return categoryMapper.toCategoryDto(category);
+    }
+
+    public CategoryResponseDto updateCategoryById(Long id, UpdateCategoryRequestDto updateDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/CategoryServiceImpl.java:47]:
```suggestion
    @Override
    public CategoryResponseDto updateCategoryById(Long id, UpdateCategoryRequestDto updateDto) {
```

File: src/main/resources/db/changelog/changes/08-create-categories-table.yaml [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+                  type: bigint
+                  autoIncrement: true
+                  constraints:
+                    primaryKey: true
+                    nullable: false
+              - column:
+                  name: name
+                  type: varchar(255)
+                  constraints:
+                    nullable: false
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/08-create-categories-table.yaml:20]: add unique

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/mate/academy/bookstore/model/User.java [hunk 13]
```diff
@@ -42,7 +43,7 @@ public class User implements UserDetails {
     private String shippingAddress;
     @Column(nullable = false)
     private boolean isDeleted = false;
-    @ManyToMany
+    @ManyToMany(fetch = FetchType.EAGER)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/User.java]: don't use Eager 
```suggestion
    @ManyToMany
```

File: src/main/java/mate/academy/bookstore/model/User.java [hunk 14]
```diff
@@ -42,7 +43,7 @@ public class User implements UserDetails {
...
     @ToStringExclude
     @EqualsAndHashCode.Exclude
     @JoinTable(name = "user_roles",
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/User.java]:
```suggestion
    @JoinTable(name = "users_roles",
```

File: src/main/java/mate/academy/bookstore/security/JwtAuthenticationFilter.java [hunk 15]
```diff
@@ -45,7 +45,7 @@ protected void doFilterInternal(
     }
 
     private String getToken(HttpServletRequest request) {
-        String bearerToken = request.getHeader(TYPE_OF_HEADER);
+        String bearerToken = request.getHeader(HttpHeaders.AUTHORIZATION);
         if (StringUtils.hasText(bearerToken) && bearerToken.startsWith(TOKEN_STARTS_WITH)) {
             return bearerToken.substring(PREFIX);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/security/JwtAuthenticationFilter.java]:
```suggestion
            return bearerToken.substring(TOKEN_STARTS_WITH.length());
```

File: src/main/java/mate/academy/bookstore/security/JwtAuthenticationFilter.java [hunk 16]
```diff
@@ -20,7 +21,6 @@
 public class JwtAuthenticationFilter extends OncePerRequestFilter {
     private static final Integer PREFIX = 7;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/security/JwtAuthenticationFilter.java]: remove

File: src/main/java/mate/academy/bookstore/security/JwtAuthenticationFilter.java [hunk 17]
```diff
@@ -20,7 +21,6 @@
...
     private static final String TOKEN_STARTS_WITH = "Bearer ";
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/security/JwtAuthenticationFilter.java]:
```suggestion
    private static final String TOKEN_HEADER = "Bearer ";
```

File: src/main/java/mate/academy/bookstore/service/UserServiceImpl.java [hunk 18]
```diff
@@ -16,6 +20,7 @@ public class UserServiceImpl implements UserService {
     private final UserRepository userRepository;
     private final UserMapper userMapper;
     private final PasswordEncoder passwordEncoder;
+    private final RoleRepository roleRepository;
 
     @Override
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/UserServiceImpl.java:26]:
```suggestion
    @Override
    @Transactional
```

File: src/main/java/mate/academy/bookstore/service/UserServiceImpl.java [hunk 19]
```diff
@@ -25,6 +30,8 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto)
         }
         User user = userMapper.toModel(requestDto);
         user.setPassword(passwordEncoder.encode(requestDto.getPassword()));
+        Role userRole = roleRepository.findByName(RoleName.ROLE_USER);
+        user.setRoles(Set.of(userRole));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/UserServiceImpl.java:36]: Have you done Shopping Cart HW ? 
It's better to create shop cart when user registers. Add method to ShopCartService and use it here

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/mate/academy/bookstore/service/UserServiceImpl.java (hunk 19 above)
Comment (Stelmaschyk) [reply src/main/java/mate/academy/bookstore/service/UserServiceImpl.java:36]: Yes, I know, I added it for only set role in table "users-roles" to test conrollers with admin role. Ok, I deleted it and implemet in next task.

File: src/main/java/mate/academy/bookstore/config/SecurityConfig.java [hunk 20]
```diff
@@ -36,13 +36,13 @@ public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Excepti
             .authorizeHttpRequests(
                 auth -> auth
                     .requestMatchers(
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/config/SecurityConfig.java:38]: that are all what you need actually 
```suggestion
                    .requestMatchers(
                                        antMatcher("/auth/**"),
                                        antMatcher("/swagger-ui/**"),
                                        antMatcher("/v3/api-docs/**"))
                                .permitAll()
```

File: src/main/java/mate/academy/bookstore/model/Book.java (hunk 7 above)
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/Book.java]: Could you please remove default fetch type? 

```suggestion
    @ManyToMany
```

File: src/main/java/mate/academy/bookstore/model/User.java [hunk 21]
```diff
@@ -42,10 +42,10 @@ public class User implements UserDetails {
     private String shippingAddress;
     @Column(nullable = false)
     private boolean isDeleted = false;
-    @ManyToMany
+    @ManyToMany //(fetch = FetchType.EAGER)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/User.java]:
```suggestion
    @ManyToMany
```

File: src/main/java/mate/academy/bookstore/service/UserServiceImpl.java (hunk 18 above)
Comment (Elena-Bruyako) [reply src/main/java/mate/academy/bookstore/service/UserServiceImpl.java:26]: I wrote [this](https://github.com/Stelmaschyk/book-store/pull/13#discussion_r1826642347) comment about shopping cart (I checked that this is a next HW) so you can ignore my comment

Setting role to user should be done in previous HWs, so please add set role

File: src/main/resources/db/changelog/changes/11-insert-categories-in-table.yaml [hunk 22]
```diff
@@ -0,0 +1,20 @@
+databaseChangeLog:
+  - changeSet:
+      id: insert-categories
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/11-insert-categories-in-table.yaml]: redundant changeset

File: src/main/resources/db/changelog/changes/12-insert-book-categories.yaml [hunk 23]
```diff
@@ -0,0 +1,35 @@
+databaseChangeLog:
+  - changeSet:
+      id: insert-books-categories
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/12-insert-book-categories.yaml]: redundant changeset

File: src/main/resources/db/changelog/changes/09-insert-books-in-table.yaml [hunk 24]
```diff
@@ -0,0 +1,59 @@
+databaseChangeLog:
+  - changeSet:
+      id: insert-book
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/09-insert-books-in-table.yaml]: redundant changeset

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/mate/academy/bookstore/model/Book.java (hunk 7 above)
Comment (Stelmaschyk) [reply src/main/java/mate/academy/bookstore/model/Book.java]: Logically))

**Review by Elena-Bruyako — APPROVED** [review]

**Pull request merged**